cirq-core/cirq/sim/simulator.py

```diff
@@ -683,6 +677,19 @@ def _base_iterator(
         Yields:
             StepResults from simulating a Moment of the Circuit.
         """
+        # In 0.16 (or 1.0) _base_iterator should be made abstract. Then _create_simulation_state,
+        # _create_act_on_args, and _core_iterator can all be removed from this class completely.
+        # (They were only required because of this implementation, which has been moved to
+        # SimulatorBase instead).
+        _compat._warn_or_error(
+            'Custom implementations of `cirq.SimulatesIntermediateState` should inherit from'
+            ' `cirq.SimulatorBase` instead, to get a richer feature set. If inheriting from'
```

Comment:
This conflicts with my understanding of `SimulatorBase`, which is that it provides shared code for Cirq-internal simulators. I'm also hesitant to suggest it to `SimulatesIntermediateState` users since `SimulatorBase` also implements `SimulatesSamples`, which external simulators may not want (e,g. because they have a more optimized sampling method).

Reply:
Fair enough, perhaps we just delete the blurb about it being deprecated and be done with it. I'm not particularly attached to this option.

I designed SimulatorBase to be a good starting point for third party simulators, but it's fine if we want to keep it internal too. 

Tangential docstring question: is Simulator actually sparse? It says it uses einsum to sparsify it, but unless I'm drastically missing something, it doesn't do that.

Reply:
Per third-party simulator base class, the one extra thing SimulatorBase requires is custom StepResult/TrialResult classes. But users shouldn't have to provide those. We can provide a ThirdPartySimulatorBase that provides default implementations of those. Then users could write a computational basis state simulator in under 40 LOC. https://github.com/quantumlib/Cirq/compare/master...daxfohl:basic-sim?expand=1

Reply:
> I designed SimulatorBase to be a good starting point for third party simulators, but it's fine if we want to keep it internal too.

I can't articulate a reason that `SimulatorBase` would need to be internal - that's just how I've always thought of it. I'm okay with presenting it as part of the Simulator API provided that we're clear about it being a "one size fits all" solution, and that the modular interfaces (`Simulates*`) are preferable for targeted-use-case simulators.

> Tangential docstring question: is Simulator actually sparse?

Not a clue. The note about einsum is probably ancient at this point - I vaguely recall discussion about trying einsum for simulation when I was only doing 20% time with the team, but nothing after that.

Reply:
Removed the SimulatorBase part of the docstring. 

Seems like we should also move the qubit_order->qubits function up into the calling function, and just pass qubits in directly here. That way implementors don't have to copy that logic. That should be backwards compatible, since qubits is a `QubitOrderOrList`. WDYT?

Also I realized we can even get rid of the need for users to create a new simulator class. This one simulates whatever state type you pass into it. https://github.com/quantumlib/Cirq/compare/master...daxfohl:basic-sim?expand=1. So down to 20 lines for a computational basis state simulator.

Reply:
`QubitOrderOrList` -> `Tuple[Qid, ...]` seems thorny since `_base_iterator` is expected to be implemented externally. In C++ it would be a `protected` method, which the subclass can freely reference and potentially pass `QubitOrder` objects, which are not qubit lists.

Reply:
I wouldn't *expect* anyone to be calling it; it's just there for `simulate_moment_steps` to call into. But I went ahead and added a warning message and conversion so it won't break if someone is calling it.